cf-oscore/src/main/java/org/eclipse/californium/oscore/OscoreOptionDecoder.java

```diff
@@ -0,0 +1,220 @@
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+/**
+ * Class for decoding the bytes of an OSCORE CoAP option.
+ * 
+ * See the structure of the option:
+ * https://datatracker.ietf.org/doc/html/rfc8613#section-6.1
+ * 
+ */
+public class OscoreOptionDecoder {
+
+	private boolean decoded;
+	private byte[] encodedBytes;
+
+	private byte[] idContext;
+	private byte[] partialIV;
+	private byte[] kid;
+
+	private int n;
+	private int k;
+	private int h;
+
+	/**
+	 * Initialize the OSCORE option with a certain array of bytes.
+	 * 
+	 * @param encodedBytes the encoded bytes of the option
+	 */
+	public OscoreOptionDecoder(byte[] encodedBytes) {
+		this.encodedBytes = encodedBytes;
+	}
+
+	/**
+	 * Set the OSCORE option to a certain array of bytes.
+	 * 
+	 * @param encodedBytes the encoded bytes of the option
+	 */
+	public void setBytes(byte[] encodedBytes) {
+		this.encodedBytes = encodedBytes;
+		decoded = false;
+	}
+
+	/**
+	 * Performs the decoding of the option and stores the resulting parameters
+	 * in this object.
+	 */
+	private void decode() {
+		byte[] total = encodedBytes;
+
+		/**
+		 * If the OSCORE option value is a zero length byte array it represents
+		 * a byte array of length 1 with a byte 0x00 See
+		 * https://tools.ietf.org/html/draft-ietf-core-object-security-16#section-2
+		 */
+		if (total.length == 0) {
+			total = new byte[] { 0x00 };
+		}
+
+		byte flagByte = total[0];
+
+		int n = flagByte & 0x07;
+		int k = flagByte & 0x08;
+		int h = flagByte & 0x10;
+
+		byte[] partialIV = null;
+		byte[] kid = null;
+		byte[] kidContext = null;
+		int index = 1;
+
+		// Parsing Partial IV
+		if (n > 0) {
+			partialIV = Arrays.copyOfRange(total, index, index + n);
+			index += n;
+		}
+
+		// Parsing KID Context
+		if (h != 0) {
+			int s = total[index];
+
+			kidContext = Arrays.copyOfRange(total, index + 1, index + 1 + s);
+
+			index += s + 1;
+		}
+
+		// Parsing KID
+		kid = Arrays.copyOfRange(total, index, total.length);
+
+		// Store parsed data in this object
+		this.n = n;
+		this.k = k;
+		this.h = h;
+		this.partialIV = partialIV;
+		this.kid = kid;
+		this.idContext = kidContext;
+	}
+
+	/**
+	 * Retrieve the ID Context
+	 * 
+	 * @return the ID Context (kid context)
+	 */
+	public byte[] getIdContext() {
+		if (!decoded) {
+			decode();
+			decoded = true;
+		}
+
+		return idContext;
+	}
+
+	/**
+	 * Retrieve the Partial IV
+	 * 
+	 * @return the Partial IV
+	 */
+	public byte[] getPartialIV() {
+		if (!decoded) {
+			decode();
+			decoded = true;
+		}
+
+		return partialIV;
+	}
+
+	/**
+	 * Retrieve the sequence number
+	 * 
+	 * @return the sequence number (based on the Partial IV)
+	 */
+	public int getSequenceNumber() {
+		if (!decoded) {
+			decode();
+			decoded = true;
+		}
+
+		if (partialIV == null) {
+			return 0;
+		}
+
+		byte[] partialIvInt = new byte[Integer.BYTES];
+		System.arraycopy(partialIV, 0, partialIvInt, 0, partialIV.length);
+
+		return ByteBuffer.wrap(partialIvInt).getInt();
```

Comment:
Did you also try to just use `ByteBuffer.wrap(partialIv).getInt()` ?

Reply:
Seems, that `partialIV` may be too short.

`Integer.BYTES` is not available in java 7 (until now, Californium is still usable with java 7, but not all features.)

I would prefer to go with:


```
byte[] partialIvInt = Arrays.copyOf(partialIV, Integer.SIZE / Byte.SIZE);
```

Reply:
Yes unfortunately partialIV is too short. I have updated the code according to your suggestion in commit 25e6209ae03b65ac32de4f6abaf00f4f03db8269